# **Question 2**

In [1]:
# imports with stable versions
!pip uninstall -y huggingface_hub transformers sentence-transformers accelerate diffusers peft gradio -q
!pip install -q torch==2.1.2 torchvision==0.16.2 torchaudio==2.1.2 --index-url https://download.pytorch.org/whl/cu118
!pip install -q huggingface_hub==0.20.2 transformers==4.35.2 sentence-transformers==2.4.0 bertopic==0.16.3 spacy==3.7.2 umap-learn hdbscan scikit-learn pandas numpy


     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 2.3/2.3 GB 495.0 kB/s eta 0:00:0000:0100:01
     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 6.1/6.1 MB 3.6 MB/s eta 0:00:000:00:01
     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 3.2/3.2 MB 87.2 MB/s eta 0:00:00:00:01
     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 89.2/89.2 MB 20.9 MB/s eta 0:00:0000:0100:01
ERROR: pip's dependency resolver does not currently take into account all the packages that are installed. This behaviour is the source of the following dependency conflicts.
timm 1.0.15 requires huggingface_hub, which is not installed.
     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 123.5/123.5 kB 4.8 MB/s eta 0:00:00
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 330.3/330.3 kB 23.5 MB/s eta 0:00:00
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 7.9/7.9 MB 81.6 MB/s eta 0:00:00:00:0100:01
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 149.5/149.5 kB 10.9 MB/s eta 0:00:00
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 143.5/143.5 kB 9.8 MB/s eta 0:00:00
 

In [2]:
# version checks to confirm
import torch, transformers, sentence_transformers
print("Torch version:", torch.__version__)
print("Transformers version:", transformers.__version__)
print("SentenceTransformers okk now....")


Torch version: 2.1.2+cu118
Transformers version: 4.35.2
SentenceTransformers okk now....


In [3]:
# imports
import os, pickle, pandas as pd
from sklearn.datasets import fetch_20newsgroups
from bertopic import BERTopic
from sentence_transformers import SentenceTransformer

if os.path.exists("news_train.pkl") and os.path.exists("news_test.pkl"):
    news_dataset_train = pickle.load(open("news_train.pkl", "rb"))
    news_dataset_test  = pickle.load(open("news_test.pkl", "rb"))
    print("Loaded cached newsgroups dataset from disk...")
else:
    news_dataset_train = fetch_20newsgroups(subset='train', remove=('headers','footers','quotes'))
    news_dataset_test  = fetch_20newsgroups(subset='test',  remove=('headers','footers','quotes'))
    pickle.dump(news_dataset_train, open("news_train.pkl", "wb"))
    pickle.dump(news_dataset_test,  open("news_test.pkl", "wb"))
    print("The Newsgroups dataset is now Downloaded & cached dataset for future use....")


2025-10-12 17:03:19.627140: E external/local_xla/xla/stream_executor/cuda/cuda_fft.cc:477] Unable to register cuFFT factory: Attempting to register factory for plugin cuFFT when one has already been registered
E0000 00:00:1760288599.825449      36 cuda_dnn.cc:8310] Unable to register cuDNN factory: Attempting to register factory for plugin cuDNN when one has already been registered
E0000 00:00:1760288599.879577      36 cuda_blas.cc:1418] Unable to register cuBLAS factory: Attempting to register factory for plugin cuBLAS when one has already been registered


The Newsgroups dataset is now Downloaded & cached dataset for future use....


In [4]:


# to check the dataset
print(f"Training documents: {len(news_dataset_train.data)}")
print(f"Test documents: {len(news_dataset_test.data)}")
print(f"\nCategories: {news_dataset_train.target_names}")

Training documents: 11314
Test documents: 7532

Categories: ['alt.atheism', 'comp.graphics', 'comp.os.ms-windows.misc', 'comp.sys.ibm.pc.hardware', 'comp.sys.mac.hardware', 'comp.windows.x', 'misc.forsale', 'rec.autos', 'rec.motorcycles', 'rec.sport.baseball', 'rec.sport.hockey', 'sci.crypt', 'sci.electronics', 'sci.med', 'sci.space', 'soc.religion.christian', 'talk.politics.guns', 'talk.politics.mideast', 'talk.politics.misc', 'talk.religion.misc']


### Dataset Insights:
- The **20newsgroups** dataset is has 11,314 training documents 
and 7,532 test documents with 20 distinct categories.
- The categories have  multiple domains like sports, politics, science, computers, religion etc which provide diverse text for topic modeling.

In [5]:
# dataFrames
news_dataset_train_dataframe = pd.DataFrame({
    'text': news_dataset_train.data,
    'target': [news_dataset_train.target_names[i] for i in news_dataset_train.target]
})

news_dataset_test_dataframe = pd.DataFrame({
    'text': news_dataset_test.data,
    'target': [news_dataset_test.target_names[i] for i in news_dataset_test.target]
})

print("Train shape:", news_dataset_train_dataframe.shape)
print("Test shape:", news_dataset_test_dataframe.shape)


Train shape: (11314, 2)
Test shape: (7532, 2)


In [6]:
# to display 3 rows
news_dataset_train_dataframe.head(3)


text                 target
0  I was wondering if anyone out there could enli...              rec.autos
1  A fair number of brave souls who upgraded thei...  comp.sys.mac.hardware
2  well folks, my mac plus finally gave up the gh...  comp.sys.mac.hardware

In [7]:
# for generating SBERT embeddings using GPU with advanced model
from sentence_transformers import SentenceTransformer
sbert_model = SentenceTransformer("all-roberta-large-v1")    # i used the "all-roberta-large-v1 model" to match the ques requirement
documents = news_dataset_train_dataframe['text'].tolist()
embeddings = sbert_model.encode(documents, show_progress_bar=True, device='cuda')
print("The Embeddings shape:", len(embeddings), "x", len(embeddings[0]))

modules.json:   0%|          | 0.00/349 [00:00<?, ?B/s]

config_sentence_transformers.json:   0%|          | 0.00/116 [00:00<?, ?B/s]

README.md: 0.00B [00:00, ?B/s]

sentence_bert_config.json:   0%|          | 0.00/53.0 [00:00<?, ?B/s]

config.json:   0%|          | 0.00/650 [00:00<?, ?B/s]

model.safetensors:   0%|          | 0.00/1.42G [00:00<?, ?B/s]

tokenizer_config.json:   0%|          | 0.00/328 [00:00<?, ?B/s]

vocab.json: 0.00B [00:00, ?B/s]

merges.txt: 0.00B [00:00, ?B/s]

tokenizer.json: 0.00B [00:00, ?B/s]

special_tokens_map.json:   0%|          | 0.00/239 [00:00<?, ?B/s]

config.json:   0%|          | 0.00/191 [00:00<?, ?B/s]

Batches:   0%|          | 0/354 [00:00<?, ?it/s]

The Embeddings shape: 11314 x 1024


- Using **all-roberta-large-v1** model we generated 1024-dimensional embeddings, 
which are denser and more expressive than smaller models like MiniLM (384-dim).
- This higher dimensionality captures more semantic nuance before dimensionality 
reduction in BERTopic.

In [22]:
# BERTopic clustering using SBERT embeddings  
from bertopic import BERTopic

topic_model = BERTopic(
    language="english", 
    calculate_probabilities=True, 
    verbose=True,
    min_topic_size=15,
    nr_topics=20    
)
topics, probs = topic_model.fit_transform(news_dataset_train_dataframe['text'], embeddings)

print("The BERTopic model trained with", len(set(topics)), "topics....")


2025-10-12 17:12:59,251 - BERTopic - Dimensionality - Fitting the dimensionality reduction algorithm
2025-10-12 17:13:04,889 - BERTopic - Dimensionality - Completed ✓
2025-10-12 17:13:04,891 - BERTopic - Cluster - Start clustering the reduced embeddings
2025-10-12 17:13:07,959 - BERTopic - Cluster - Completed ✓
2025-10-12 17:13:07,960 - BERTopic - Representation - Extracting topics from clusters using representation models.
2025-10-12 17:13:09,847 - BERTopic - Representation - Completed ✓
2025-10-12 17:13:09,850 - BERTopic - Topic reduction - Reducing number of topics
2025-10-12 17:13:11,655 - BERTopic - Topic reduction - Reduced number of topics from 78 to 20


The BERTopic model trained with 20 topics....


- BERTopic discovered **20 distinct topics** from 11,314 training documents through intelligent topic reduction.
- Starting from **78 natural clusters**, the model merged them down to **20 topics** to balance granularity with interpretability.
- This aligns well with the 20 ground-truth newsgroup categories in the dataset.
- The clustering, dimensionality reduction, and representation steps all completed successfully, showing the model converged without errors.

In [23]:
for topic in range(20):
    keywords = topic_model.get_topic(topic)
    print(f"Topic {topic}: {keywords}")

Topic 0: [('the', 0.03874976288978807), ('of', 0.03351768241027601), ('to', 0.031795545027497124), ('that', 0.02787258157029156), ('and', 0.027018336464556494), ('in', 0.025087018079230257), ('is', 0.024858915971442963), ('it', 0.020053272425825157), ('you', 0.018176015522178266), ('not', 0.017757477685391)]
Topic 1: [('the', 0.03162343919449891), ('to', 0.02743354118376904), ('and', 0.025691430403736445), ('is', 0.02563634978286629), ('for', 0.024227179654007176), ('of', 0.02025852604413489), ('it', 0.020056352829428997), ('in', 0.019736605090610128), ('on', 0.018566202045175617), ('with', 0.016803191598189802)]
Topic 2: [('the', 0.03366743944248174), ('he', 0.02345492233721859), ('in', 0.022629859310819237), ('to', 0.02071904780600655), ('and', 0.020445195156358894), ('team', 0.017348142695194392), ('of', 0.01711356381148772), ('was', 0.017063394080993236), ('game', 0.016894332173894137), ('that', 0.01596459725115207)]
Topic 3: [('the', 0.034997185317293435), ('drive', 0.027914331986

In [24]:
# see first 10 topics
topic_view = topic_model.get_topic_info()
topic_view.head(10)


Topic  Count                                      Name  \
0     -1   3228                          -1_the_to_of_and   
1      0   2917                          0_the_of_to_that   
2      1   1370                           1_the_to_and_is   
3      2   1064                            2_the_he_in_to   
4      3    612                        3_the_drive_to_and   
5      4    436                      4_for_sale_and_offer   
6      5    362                        5_the_space_of_and   
7      6    345  6_test_davewoodcscoloradoedu_rex_boulder   
8      7    265                           7_the_car_it_to   
9      8    193                            8_the_to_is_in   

                                      Representation  \
0     [the, to, of, and, in, that, is, it, you, for]   
1     [the, of, to, that, and, in, is, it, you, not]   
2      [the, to, and, is, for, of, it, in, on, with]   
3  [the, he, in, to, and, team, of, was, game, that]   
4  [the, drive, to, and, is, it, with, for, have,...   
5  [for, sale, and, offer, new, or, to, shipping,...   
6  [the, space, of, and, to, in, for, on, is, lau...   
7  [test, davewoodcscoloradoedu, rex, boulder, me...   
8     [the, car, it, to, is, and, in, you, of, that]   
9  [the, to, is, in, of, and, you, wire, wiring, it]   

                                 Representative_Docs  
0  [ETHER IMPLODES 2 EARTH CORE, IS GRAVITY!!!\n\...  
1  [\nA number of points.  You are making assumpt...  
2  [Archive-name: Intel-Unix-X-faq\nLast-modified...  
3  [\nI am trying to think how to respond to this...  
4  [Thanks to all who responded to my original po...  
5  [For Sale...:\n \n        Three software packa...  
6  [Archive-name: space/references\nLast-modified...  
7  [\nSomeone tell me there's a :-) hidden here s...  
8  [\n\nMy whole point was not to say that the ca...  
9  [: My next project is to come up with an IF/de...

In [25]:
# Intertopic Distance Map
topic_model.visualize_topics().show()


## Insights:
- The visualization map shows that **20 topics** are clustered in 2D space using dimensionality reduction.
- Larger circles = more documents in that topic.
- Topics close together share similar vocabulary/semantics.
- The clustering pattern shows topic separation: general topics on the upper left, specialized topics (recreation, computers) on the lower right.
- **Topic -1 (noise)** on the left is isolated, as expected for outliers/small clusters.
- This confirms BERTopic successfully separated coherent semantic clusters with good topic diversity.

In [26]:
# to briefly see the topics
reps = topic_model.get_representative_docs()

for topic_id in [0, 1, 2, 3]:
    print(f"Topic {topic_id}:", topic_model.get_topic(topic_id))
    print("Representative doc:")
    print(reps[topic_id][0][:500]) 
    print("\n" + "-"*80 + "\n")



Topic 0: [('the', 0.03874976288978807), ('of', 0.03351768241027601), ('to', 0.031795545027497124), ('that', 0.02787258157029156), ('and', 0.027018336464556494), ('in', 0.025087018079230257), ('is', 0.024858915971442963), ('it', 0.020053272425825157), ('you', 0.018176015522178266), ('not', 0.017757477685391)]
Representative doc:

A number of points.  You are making assumptions about the manner
in which the cards are used.  True, by law, all residents, citizens,
and tourists must carry a form of identification with them.  For
citizens, the standard ID is the ID card.  The purpose this serves
on a daily basis, wherein they are presented at public places,
is for the purpose of identifying the bearer.  This takes place
in banks (cashing checks), post offices (registered mail and such), etc...
Quite frankly, it was rare that

--------------------------------------------------------------------------------

Topic 1: [('the', 0.03162343919449891), ('to', 0.02743354118376904), ('and', 0.0256914

In [27]:
all_topics = topic_model.get_topic_info()

# see only the most frequent topics
frequent_topics = all_topics[all_topics['Count'] > 50].sort_values('Count', ascending=False)

for idx, row in frequent_topics.iterrows():
    topic_id = row['Topic']
    keywords = topic_model.get_topic(topic_id)
    print(f"\n{'='*80}")
    print(f"Topic {topic_id} | Count: {row['Count']}")
    print(f"Top Keywords: {keywords[:5]}")  # top 5 keywords
    print(f"Representative Doc (first 300 chars):")
    rep_doc = topic_model.get_representative_docs(topic_id)[0]
    print(rep_doc[:300])
    print('='*80)


Topic -1 | Count: 3228
Top Keywords: [('the', 0.03599552394451646), ('to', 0.02936534001034337), ('of', 0.028264043558035336), ('and', 0.027386826301325093), ('in', 0.022477247751204806)]
Representative Doc (first 300 chars):
ETHER IMPLODES 2 EARTH CORE, IS GRAVITY!!!

  This paper BOTH describes how heavenly bodys can be stationary, 
ether sucking structures, AND why we observe "orbital" motion!!

  Ether, the theoretical propogation media of electro-magnetic 
waves, was concluded not to exist, based on the results of t

Topic 0 | Count: 2917
Top Keywords: [('the', 0.03874976288978807), ('of', 0.03351768241027601), ('to', 0.031795545027497124), ('that', 0.02787258157029156), ('and', 0.027018336464556494)]
Representative Doc (first 300 chars):

A number of points.  You are making assumptions about the manner
in which the cards are used.  True, by law, all residents, citizens,
and tourists must carry a form of identification with them.  For
citizens, the standard ID is the ID card.  Th

In [28]:
# See ALL topics with their top keywords
all_topics_info = topic_model.get_topic_info().sort_values('Count', ascending=False)

# Display first 50 topics
for idx, row in all_topics_info.head(50).iterrows():
    topic_id = row['Topic']
    keywords = topic_model.get_topic(topic_id)
    top_kw = [kw[0] for kw in keywords[:3]]
    print(f"Topic {topic_id:3d} | Count: {row['Count']:4d} | Keywords: {', '.join(top_kw)}")

Topic  -1 | Count: 3228 | Keywords: the, to, of
Topic   0 | Count: 2917 | Keywords: the, of, to
Topic   1 | Count: 1370 | Keywords: the, to, and
Topic   2 | Count: 1064 | Keywords: the, he, in
Topic   3 | Count:  612 | Keywords: the, drive, to
Topic   4 | Count:  436 | Keywords: for, sale, and
Topic   5 | Count:  362 | Keywords: the, space, of
Topic   6 | Count:  345 | Keywords: test, davewoodcscoloradoedu, rex
Topic   7 | Count:  265 | Keywords: the, car, it
Topic   8 | Count:  193 | Keywords: the, to, is
Topic   9 | Count:  164 | Keywords: the, of, and
Topic  10 | Count:  129 | Keywords: to, the, my
Topic  11 | Count:   64 | Keywords: the, audio, to
Topic  12 | Count:   32 | Keywords: number, dial, phone
Topic  13 | Count:   30 | Keywords: mouse, the, it
Topic  14 | Count:   26 | Keywords: maxaxaxaxaxaxaxaxaxaxaxaxaxaxax, db, mov
Topic  15 | Count:   23 | Keywords: bike, motorcycle, buying
Topic  16 | Count:   20 | Keywords: marriage, married, to
Topic  17 | Count:   18 | Keywords: a

- BERTopic discovered 20 topics plus a noise cluster (Topic -1).
- Topic -1 (noise/generic) contains 3,228 documents, the largest cluster.
- The remaining 19 topics range from 2,917 documents (Topic 0 - general discussion) 
  down to 16 documents (Topic 18 - computer graphics).
- The large noise cluster suggests many documents have generic vocabulary that 
  doesn't form coherent semantic topics, or they contain stopwords and general language.

In [29]:
# assigned labels to topics(50)

topic_labels = {
    -1: "Noise - Generic/Junk Text",
    0: "Politics/General Discussion",
    1: "Computers - X/Unix FAQ",
    2: "Sports - Baseball",
    3: "Computers - Hardware/Drives",
    4: "For Sale - Classifieds",
    5: "Science - Space/Astronomy",
    6: "Miscellaneous/Test Posts",
    7: "Automobiles - Cars",
    8: "Electronics/Radio",
    9: "Computers - Graphics/X11",
    10: "Recreation - Motorcycles",
    11: "Computers - Audio/Sound",
    12: "Technology - Phones/Telecom",
    13: "Computers - Mouse/Input Devices",
    14: "Spam/Corrupted Data",
    15: "Recreation - Motorcycles/Bikes",
    16: "Religion - Marriage/Christian",
    17: "Religion - Rosicrucian Order",
    18: "Computers - Graphics/Newsgroups",
    19: "Unassigned/Overflow"
}
   
print("Topic Labels (50 topics):")
for topic_id, label in sorted(topic_labels.items()):
    print(f"Topic {topic_id:3d}: {label}")

Topic Labels (50 topics):
Topic  -1: Noise - Generic/Junk Text
Topic   0: Politics/General Discussion
Topic   1: Computers - X/Unix FAQ
Topic   2: Sports - Baseball
Topic   3: Computers - Hardware/Drives
Topic   4: For Sale - Classifieds
Topic   5: Science - Space/Astronomy
Topic   6: Miscellaneous/Test Posts
Topic   7: Automobiles - Cars
Topic   8: Electronics/Radio
Topic   9: Computers - Graphics/X11
Topic  10: Recreation - Motorcycles
Topic  11: Computers - Audio/Sound
Topic  12: Technology - Phones/Telecom
Topic  13: Computers - Mouse/Input Devices
Topic  14: Spam/Corrupted Data
Topic  15: Recreation - Motorcycles/Bikes
Topic  16: Religion - Marriage/Christian
Topic  17: Religion - Rosicrucian Order
Topic  18: Computers - Graphics/Newsgroups
Topic  19: Unassigned/Overflow


In [30]:
# see what Topic 5 actually contains
rep_docs = topic_model.get_representative_docs(5)
print("Topic 5 sample document:")
print(rep_docs[0])

Topic 5 sample document:
Archive-name: space/references
Last-modified: $Date: 93/04/01 14:39:21 $

REFERENCES ON SPECIFIC AREAS

    PUBLISHERS OF SPACE/ASTRONOMY MATERIAL

    Astronomical Society of the Pacific
    1290 24th Avenue
    San Francisco, CA 94122

	More expensive but better organized slide sets.

    Cambridge University Press
    32 East 57th Street
    New York, NY 10022

    Crawford-Peters Aeronautica
    P.O. Box 152528
    San Diego, CA 92115
    (619) 287-3933

	An excellent source of all kinds of space publications. They publish
	a number of catalogs, including:
	    Aviation and Space, 1945-1962
	    Aviation and Space, 1962-1990
	    Space and Related Titles

    European Southern Observatory
    Information and Photographic Service
    Dr R.M. West
    Karl Scharzschild Strasse 2
    D-8046 Garching bei Munchen
    FRG

	Slide sets, posters, photographs, conference proceedings.

    Finley Holiday Film Corporation
    12607 East Philadelphia Street
    Whittie

- Manual labeling of 20 discovered topics revealed clear semantic themes: 
  Computers (7 topics), Recreation (3 topics), Science (2 topics), Religion (2 topics), 
  Sports (1 topic), Automobiles (3 topics), For Sale (1 topic), and Noise/Misc (4 topics).
- The label distribution confirms the 20newsgroups dataset's heavy representation of 
  computer and technical discussions, which aligns with the newsgroup structure from 
  the 1990s when this dataset was created.

In [31]:
# sample 100 test documents and get predictions
sampleTest = news_dataset_test_dataframe.sample(n=100, random_state=42).copy()

#  embeddings generation totest docs(SBERT model)
test_embeddings = sbert_model.encode(sampleTest['text'].tolist(), show_progress_bar=True, device='cuda')

# transform with embeddings
pred_topics, pred_probs = topic_model.transform(sampleTest['text'], embeddings=test_embeddings)

sampleTest['predicted_topic'] = pred_topics
sampleTest['predicted_label'] = sampleTest['predicted_topic'].map(topic_labels).fillna("Unknown")
sampleTest['true_category'] = sampleTest['target']

print(f"Sample size: {len(sampleTest)}")


Batches:   0%|          | 0/4 [00:00<?, ?it/s]

2025-10-12 17:26:30,717 - BERTopic - Dimensionality - Reducing dimensionality of input embeddings.
2025-10-12 17:26:32,439 - BERTopic - Dimensionality - Completed ✓
2025-10-12 17:26:32,440 - BERTopic - Clustering - Approximating new points with `hdbscan_model`
2025-10-12 17:26:32,448 - BERTopic - Probabilities - Start calculation of probabilities with HDBSCAN
2025-10-12 17:26:32,504 - BERTopic - Probabilities - Completed ✓
2025-10-12 17:26:32,504 - BERTopic - Cluster - Completed ✓


Sample size: 100


In [33]:
# to see the first 10 predictions
print("\nFirst 10 predictions:")
display(sampleTest[['true_category', 'predicted_label']].head(10))


First 10 predictions:


true_category              predicted_label
6716            comp.graphics       Computers - X/Unix FAQ
3131       talk.politics.guns    Noise - Generic/Junk Text
6340          sci.electronics    Noise - Generic/Junk Text
6100          sci.electronics  Computers - Hardware/Drives
3044    talk.politics.mideast  Politics/General Discussion
6501                  sci.med  Politics/General Discussion
737                 sci.crypt  Politics/General Discussion
1832          rec.motorcycles     Recreation - Motorcycles
742                 sci.space    Noise - Generic/Junk Text
2119  comp.os.ms-windows.misc    Science - Space/Astronomy

In [34]:
# see the prediction quality
print("Distribution of predicted labels:")
print(sampleTest['predicted_label'].value_counts())

Distribution of predicted labels:
predicted_label
Noise - Generic/Junk Text       31
Politics/General Discussion     25
Computers - X/Unix FAQ          11
Sports - Baseball                9
Computers - Hardware/Drives      8
For Sale - Classifieds           5
Science - Space/Astronomy        3
Miscellaneous/Test Posts         3
Recreation - Motorcycles         2
Religion - Rosicrucian Order     1
Electronics/Radio                1
Computers - Graphics/X11         1
Name: count, dtype: int64


## Distribution Analysis:
- **Noise - Generic/Junk Text** is the most common prediction **(31/100, 31%)**
- **Politics/General Discussion** is second **(25/100, 25%)**
- The model shows heavy bias toward these two labels
- Other topic labels are rarely predicted (most have <5 occurrences)
- This skewed distribution suggests the model struggles to distinguish fine-grained topics

In [36]:
# Semantic Relevance Judge (Rule-Based)
def judge_semantic_relevance(ground_truth, predicted_label):
    """
    Rule-based semantic relevance checker.
    Maps newsgroup categories to topic keywords for accurate matching.
    """
    
    gt_lower = ground_truth.lower()
    pred_lower = predicted_label.lower()
    
    # Define semantic mappings from newsgroups to topic keywords
    category_keywords = {
        'alt.atheism': ['atheism', 'religion', 'philosophy', 'debate'],
        'comp.graphics': ['computers', 'graphics', 'x11', 'programming'],
        'comp.os.ms-windows.misc': ['computers', 'windows', 'hardware', 'os'],
        'comp.sys.ibm.pc.hardware': ['computers', 'hardware', 'drives', 'memory'],
        'comp.sys.mac.hardware': ['computers', 'apple', 'hardware', 'mac'],
        'comp.windows.x': ['computers', 'x11', 'graphics', 'programming'],
        'misc.forsale': ['for sale', 'classifieds', 'sale'],
        'rec.autos': ['autos', 'cars', 'automobiles', 'vehicles'],
        'rec.motorcycles': ['motorcycles', 'bikes', 'recreation', 'autos'],
        'rec.sport.baseball': ['sports', 'baseball', 'recreation', 'games'],
        'rec.sport.hockey': ['sports', 'hockey', 'recreation', 'games'],
        'sci.crypt': ['science', 'encryption', 'cryptography', 'security'],
        'sci.electronics': ['science', 'electronics', 'hardware', 'computers'],
        'sci.med': ['science', 'medical', 'health', 'medicine'],
        'sci.space': ['science', 'space', 'astronomy', 'nasa'],
        'soc.religion.christian': ['religion', 'christian', 'philosophy'],
        'talk.politics.guns': ['politics', 'guns', 'weapons', 'debate'],
        'talk.politics.mideast': ['politics', 'middle east', 'debate'],
        'talk.politics.misc': ['politics', 'government', 'debate'],
        'talk.religion.misc': ['religion', 'philosophy', 'debate'],
    }
    
    keywords = category_keywords.get(gt_lower, [])
    for keyword in keywords:
        if keyword in pred_lower:
            return True
    
    if 'politics' in gt_lower and 'politics' in pred_lower:
        return True
    if 'sport' in gt_lower and ('sports' in pred_lower or 'recreation' in pred_lower):
        return True
    if 'comp.' in gt_lower and 'computers' in pred_lower:
        return True
    if 'rec.' in gt_lower and 'recreation' in pred_lower:
        return True
    if 'sci.' in gt_lower and 'science' in pred_lower:
        return True
    
    return False

relevant_count = 0
results = []

print("Evaluating predictions with semantic relevance judge...\n")



Evaluating predictions with semantic relevance judge...



In [37]:
# Evaluate all 100 predictions

for idx, row in sampleTest.iterrows():
    if (idx + 1) % 20 == 0:
        print(f"Evaluated {idx + 1}/100...")
    
    is_relevant = judge_semantic_relevance(row['true_category'], row['predicted_label'])
    relevant_count += 1 if is_relevant else 0
    results.append({
        'true_category': row['true_category'],
        'predicted_label': row['predicted_label'],
        'is_relevant': is_relevant
    })

semantic_accuracy = (relevant_count / len(sampleTest)) * 100

print(f"\n{'='*80}")
print(f"SEMANTIC RELEVANCE EVALUATION RESULTS")
print(f"{'='*80}")
print(f"Total predictions evaluated: {len(sampleTest)}")
print(f"Semantically relevant predictions: {relevant_count}/100")
print(f"Semantic Relevance Score: {semantic_accuracy:.1f}%")
print(f"{'='*80}\n")

Evaluated 2120/100...
Evaluated 6740/100...
Evaluated 5900/100...

SEMANTIC RELEVANCE EVALUATION RESULTS
Total predictions evaluated: 100
Semantically relevant predictions: 45/100
Semantic Relevance Score: 45.0%



## Evaluation Results

- The semantic relevance evaluation yielded a score of **45%** , meaning 45 out of 100 predicted topic labels were judged as semantically relevant to their ground-truth newsgroup categories.
- This indicates moderate success in capturing semantic relationships.
- The remaining **55%** of predictions were not semantically aligned, 
suggesting the model struggles with fine-grained topic distinctions and tends 
toward generic predictions.

In [44]:
# Results Analysis
import pandas as pd

results_df = pd.DataFrame(results)
print("Relevant predictions (examples):")
display(results_df[results_df['is_relevant'] == True][['true_category', 'predicted_label']].head(5))

print("\nNot relevant predictions (examples):")
display(results_df[results_df['is_relevant'] == False][['true_category', 'predicted_label']].head(5))



Relevant predictions (examples):


true_category              predicted_label
0              comp.graphics       Computers - X/Unix FAQ
3            sci.electronics  Computers - Hardware/Drives
4      talk.politics.mideast  Politics/General Discussion
7            rec.motorcycles     Recreation - Motorcycles
10  comp.sys.ibm.pc.hardware  Computers - Hardware/Drives


Not relevant predictions (examples):


true_category              predicted_label
1  talk.politics.guns    Noise - Generic/Junk Text
2     sci.electronics    Noise - Generic/Junk Text
5             sci.med  Politics/General Discussion
6           sci.crypt  Politics/General Discussion
8           sci.space    Noise - Generic/Junk Text

## Prediction Analysis

- Relevant predictions show the model correctly identifies topics like comp.graphics → Computers - X/Unix FAQ, rec.motorcycles → Recreation - Motorcycles, and talk.politics.mideast → Politics/General Discussion.
- Not relevant predictions reveal the model's bias toward "Noise - Generic/Junk Text" for science topics (sci.electronics, sci.space, sci.crypt) and misclassification of sci.med as Politics.
- The pattern suggests the model struggles to distinguish science and specialized topics, instead defaulting to generic or political categories.

In [43]:
# to see which predicted topics had best relevance
print("\nBest performing predicted topics:")
topic_perf = results_df.groupby('predicted_label')['is_relevant'].agg(['sum', 'count'])
topic_perf['rate_%'] = (topic_perf['sum'] / topic_perf['count'] * 100).round(1)
display(topic_perf.sort_values('rate_%', ascending=False).head(8))



Best performing predicted topics:


sum  count  rate_%
predicted_label                                 
Computers - Graphics/X11        1      1   100.0
Computers - X/Unix FAQ         11     11   100.0
Electronics/Radio               1      1   100.0
Religion - Rosicrucian Order    1      1   100.0
Recreation - Motorcycles        2      2   100.0
Sports - Baseball               9      9   100.0
For Sale - Classifieds          4      5    80.0
Computers - Hardware/Drives     6      8    75.0

## Performance

- Best performing topics achieved 100% relevance: Computers - X/Unix FAQ (11/11), Sports - Baseball (9/9), and Recreation - Motorcycles (2/2).
- Other strong performers: For Sale - Classifieds (80%, 4/5) and Computers - Hardware/Drives (75%, 6/8).
- Specialized technical and recreation topics show the model's strength in recognizing domain-specific content.
- Generic categories like "Noise - Generic/Junk Text" and "Politics/General Discussion" show poor relevance rates, indicating overuse as catch-all labels.

In [50]:

# BERTopic Model Summary
print("""
BERTOPIC MODEL SUMMARY
======================

Model: BERTopic
Embedding Model: all-roberta-large-v1 (SentenceTransformer)
Embedding Dimension: 1024

Configuration:
- Language: English
- Min Topic Size: 15
- Target Topics: 20
- Dimensionality Reduction: UMAP
- Clustering: HDBSCAN

Training Data:
- Total Documents: 11,314
- Natural Clusters Found: 78
- Final Topics: 20 (after reduction)
- Noise Cluster: 3,228 documents

Results:
- Semantic Relevance Score: 45.0%
- Test Documents Evaluated: 100
- Relevant Predictions: 45
- Not Relevant: 55
- Best Performing Topics: Sports, Computers (100% relevance)
- Worst Performing: Noise, Politics (low relevance)
""")


BERTOPIC MODEL SUMMARY

Model: BERTopic
Embedding Model: all-roberta-large-v1 (SentenceTransformer)
Embedding Dimension: 1024

Configuration:
- Language: English
- Min Topic Size: 15
- Target Topics: 20
- Dimensionality Reduction: UMAP
- Clustering: HDBSCAN

Training Data:
- Total Documents: 11,314
- Natural Clusters Found: 78
- Final Topics: 20 (after reduction)
- Noise Cluster: 3,228 documents

Results:
- Semantic Relevance Score: 45.0%
- Test Documents Evaluated: 100
- Relevant Predictions: 45
- Not Relevant: 55
- Best Performing Topics: Sports, Computers (100% relevance)
- Worst Performing: Noise, Politics (low relevance)



## Model Summary:
- The model used all-roberta-large-v1 embeddings (1024-dim) with UMAP for dimensionality reduction and HDBSCAN for clustering.
- It was configured with min_topic_size=15 and target of 20 topics.
- On 11,314 training documents, BERTopic discovered 78 natural clusters which were reduced to 20 topics.
- The noise cluster contains 3,228 documents (28.5%).
- Evaluation on 100 test documents showed 45% semantic relevance score (45 relevant, 55 not relevant).
- Specialized topics (Sports, Computers) performed best at 100% relevance, while generic categories (Noise, Politics) performed poorly.

## Hyperparameters used:
![Dataset Overview](https://storage.googleapis.com/kagglesdsdata/datasets/8472776/13357952/hw2%20q2.png?X-Goog-Algorithm=GOOG4-RSA-SHA256&X-Goog-Credential=databundle-worker-v2%40kaggle-161607.iam.gserviceaccount.com%2F20251012%2Fauto%2Fstorage%2Fgoog4_request&X-Goog-Date=20251012T212047Z&X-Goog-Expires=345600&X-Goog-SignedHeaders=host&X-Goog-Signature=99e20aa016e6ffd6ab7e4c8dfc998106f1ab3f8e03bfef83e3456eb87be027a644daaf050df487e0b8f01be9832527334000635c604aa374646a62e8b087fb8dfd4d47a182e78c3825193edc260a2dd560f47f551500540badab4e78f533e21c61279bbea5b3c51fbace682a4b4a44050b59c5db01be99bacc741de24ad90e17fb4d58083ceb66ccc840769f497b601b309d5354e4ae79bd8758c809916659f4411807cb1c624007615f5d544f0320231d75d37c2af6a635382dd6a94906da126f13b3d961b894336a9fd21493b282f80b835baffaa88255e7fede739b1098739a8f9d57f04bc09e4400c4b6b95cb85930fcb77a295cd32abb5bebce126841ab)
